File: src/snapkitty/ltms/ltms.pl
%% SKC-LISP: Layered Truth Maintenance System (LTMS)
%% Knowledge layer: Prolog symbolic reasoning engine
%% Author: Ahmad Parr (ahmedparr93@gmail.com)
%% Integration: Snapkitty LISP Bridge knowledge engine

:- dynamic fact/5.        % fact(Value, Source, Timestamp, Confidence, Priority)
:- dynamic rule/4.        % rule(Head, Body, Module, Priority)
:- dynamic concept/2.     % concept(Name, Senses)
:- dynamic sense/4.       % sense(Concept, Gloss, ContextPred, Confidence)
:- dynamic belief_base/1. % belief_base(Timestamp) for history
:- dynamic module_stats/3. % module_stats(Module, RuleCount, LastUpdated)

%% ============================================================================
%% 1. CONFLICT RESOLUTION
%% ============================================================================
%% When multiple facts claim the same value, pick the winner by:
%% Priority > Confidence > Source recency

resolve_conflict(Value, Winner) :-
    findall(f(Value,S,T,C,P), fact(Value,S,T,C,P), Candidates),
    (Candidates = []
      -> fail
      ;  predsort(compare_facts, Candidates, Sorted),
         Sorted = [Winner|_]).

%% Comparison: higher priority wins, then higher confidence
compare_facts(Order, f(_,_,_,C1,P1), f(_,_,_,C2,P2)) :-
    ( P1 > P2 -> Order = (<)
    ; P1 < P2 -> Order = (>)
    ; C1 > C2 -> Order = (<)
    ; C1 < C2 -> Order = (>)
    ; Order = (=) ).

%% Get the winning value
get_belief(Value, WinningFact) :-
    resolve_conflict(Value, WinningFact).

%% ============================================================================
%% 2. OUTDATED DETECTION
%% ============================================================================
%% Exponential decay: confidence decays over time
%% Decay function: Conf(t) = Conf(0) * exp(-0.0001 * age_in_ms)

is_outdated(fact(_,_,Created,Conf,_), Now) :-
    Decay is Conf * exp(-0.0001 * (Now - Created)),
    Decay < 0.15.

prune_outdated_facts(Now) :-
    findall(f(V,S,T,C,P), fact(V,S,T,C,P), AllFacts),
    forall(member(F, AllFacts),
           (is_outdated(F, Now) -> retract(F) ; true)).

%% Half-life: time for confidence to drop to 50%
compute_half_life(HalfLife) :-
    HalfLife is log(2.0) / 0.0001.  % ~6931 milliseconds = 6.9 seconds

%% ============================================================================
%% 3. AMBIGUOUS CONCEPTS
%% ============================================================================
%% Concepts have multiple senses, each with a context predicate

register_concept(Name, SenseGlosses) :-
    findall(sense(Name, Gloss, _, 1.0), member(Gloss, SenseGlosses), Senses),
    assertz(concept(Name, Senses)).

disambiguate(Concept, Context, Sense) :-
    concept(Concept, Senses),
    member(sense(Concept, Gloss, Pred, Conf), Senses),
    (var(Pred) ; call(Pred, Context)),
    Sense = sense(Concept, Gloss, Pred, Conf).

%% Pick best sense by confidence (or first if all equal)
best_sense(Concept, Context, BestSense) :-
    findall(Sense, disambiguate(Concept, Context, Sense), Candidates),
    (Candidates = []
      -> fail
      ;  sort(Candidates, [BestSense|_])).

%% ============================================================================
%% 4. MAINTAINABILITY GUARD
%% ============================================================================
%% Hard limit: no module can have > 80 rules
%% Prevents knowledge explosion and maintains readable rule sets

assert_rule(Head, Body, Module, Priority) :-
    findall(_, rule(_,_,Module,_), Rules),
    length(Rules, N),
    (N >= 80
      -> throw(error(module_overflow,
                     context(assert_rule/4,
                             "Module " + Module + " has reached rule limit")))
      ;  assertz(rule(Head, Body, Module, Priority)),
         update_module_stats(Module)).

update_module_stats(Module) :-
    (retract(module_stats(Module, _, _)) ; true),
    findall(_, rule(_,_,Module,_), Rules),
    length(Rules, Count),
    get_time(Now),
    assertz(module_stats(Module, Count, Now)).

%% Get module health (rules / max_rules)
module_complexity(Module, Percentage) :-
    module_stats(Module, Count, _),
    Percentage is (Count / 80.0) * 100.

%% Suggest refactoring when > 70% full
suggest_refactor(Module) :-
    module_complexity(Module, P),
    P > 70.0.

%% ============================================================================
%% 5. HYBRID / NON-RULE KNOWLEDGE
%% ============================================================================
%% Fallback to embeddings when symbolic reasoning fails

hybrid_prove(Goal, Result, Confidence) :-
    ( prove_symbolic(Goal, Result)
    -> Confidence = 1.0
    ;  embedding_retrieve(Goal, Result, Confidence)
    ).

prove_symbolic(Goal, Result) :-
    clause(Goal, Body),
    call(Body),
    Result = Goal.

%% Stub for embedding retrieval (connects to vector DB)
embedding_retrieve(Goal, Result, Confidence) :-
    % In production: query Qdrant/WORM ledger for similar embeddings
    % For now: fail gracefully
    fail.

%% ============================================================================
%% ASSERTION & QUERY API
%% ============================================================================

%% Assert a new fact
assert_fact(Value, Source, Confidence, Priority) :-
    get_time(Now),
    Timestamp is floor(Now * 1000),  % milliseconds
    assertz(fact(Value, Source, Timestamp, Confidence, Priority)).

%% Query belief about a value (returns winning fact + confidence)
query(Value, Confidence) :-
    resolve_conflict(Value, f(Value, _, _, Confidence, _)).

%% All known senses of a concept
all_senses(Concept, AllSenses) :-
    concept(Concept, AllSenses).

%% Concept subsumption: is Specific a kind of General?
is_a(Specific, General) :-
    fact(is_a(Specific, General), _, _, _, _).

%% ============================================================================
%% DEBUGGING & INSPECTION
%% ============================================================================

%% List all facts
list_facts :-
    findall(f(V,S,T,C,P), fact(V,S,T,C,P), Facts),
    forall(member(F, Facts), writeln(F)).

%% List all rules by module
list_rules_in_module(Module) :-
    findall(rule(H,B,M,Pr), rule(H,B,Module,Pr), Rules),
    forall(member(R, Rules), writeln(R)).

%% List all concepts
list_concepts :-
    findall(C, concept(C, _), Concepts),
    forall(member(C, Concepts), writeln(C)).

%% Module report
module_report :-
    findall(m(M,Count,Updated), module_stats(M,Count,Updated), Stats),
    forall(member(m(Mod,Count,_), Stats),
           format('~w: ~w/80 rules~n', [Mod, Count])).

%% ============================================================================
%% TESTS
%% ============================================================================

:- initialization(run_ltms_tests).

run_ltms_tests :-
    writeln('=== LTMS Tests ==='),

    % Test 1: Conflict resolution
    writeln('Test 1: Conflict resolution'),
    assert_fact(color(car), source_a, 0.8, 10),
    assert_fact(color(car), source_b, 0.9, 5),
    resolve_conflict(color(car), Winner),
    writeln(Winner), nl,

    % Test 2: Outdated detection
    writeln('Test 2: Outdated detection'),
    get_time(Now),
    OldTime is (Now * 1000) - 10000,
    (is_outdated(fact(old_value, old_source, OldTime, 0.5, 1), Now * 1000)
      -> writeln('✓ Old fact marked as outdated')
      ;  writeln('✗ Old fact not detected')), nl,

    % Test 3: Module limit
    writeln('Test 3: Module complexity guard'),
    (module_complexity(main, P), format('Main module: ~1f% full~n', [P])
      ;  writeln('Main module not yet used')), nl,

    writeln('=== LTMS Tests Complete ===').

File: src/snapkitty/ltms/LTMS.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}

-- SKC-LISP: Layered Truth Maintenance System (LTMS) — Haskell Implementation
-- Knowledge layer: Type-safe symbolic reasoning
-- Author: Ahmad Parr (ahmedparr93@gmail.com)
-- Integration: Snapkitty LISP Bridge knowledge engine

module Snapkitty.LTMS.LTMS where

import Data.Time
import Data.List (sortBy, sortOn)
import Data.Ord (Down(..))
import Data.Map (Map)
import qualified Data.Map as Map
import Data.Set (Set)
import qualified Data.Set as Set
import Control.Monad (when)
import GHC.Generics
import Text.Printf

-- ============================================================================
-- DATA STRUCTURES
-- ============================================================================

-- A Fact: asserts a value from a source with confidence + priority
data Fact = Fact
  { value :: String
  , source :: String
  , timestamp :: UTCTime
  , confidence :: Double       -- 0.0 to 1.0
  , priority :: Int            -- higher = wins conflicts
  } deriving (Show, Eq, Generic)

-- A Sense: one interpretation of a concept
data Sense = Sense
  { gloss :: String                           -- human-readable meaning
  , contextPred :: Maybe (Context -> Bool)    -- predicate: "applies in this context?"
  , confidence :: Double                      -- how confident we are in this sense
  } deriving (Show, Eq)

-- A Concept: word/term with multiple possible meanings
data Concept = Concept
  { name :: String
  , senses :: [Sense]
  } deriving (Show, Eq)

-- Context: a key-value map for disambiguating senses
type Context = Map.Map String String

-- A Rule: head :- body within a module
data Rule = Rule
  { ruleHead :: String
  , ruleBody :: String
  , ruleModule :: String
  , rulePriority :: Int
  } deriving (Show, Eq)

-- Module statistics: for tracking complexity
data ModuleStats = ModuleStats
  { module :: String
  , ruleCount :: Int
  , maxRules :: Int        -- hard limit (80)
  , lastUpdated :: UTCTime
  } deriving (Show, Eq)

-- Result of hybrid query: symbolic proof vs embedding fallback
data HybridResult a
  = Symbolic a Double      -- proof + confidence (1.0 for logical proof)
  | Embedding a Double     -- embedding retrieval + confidence
  | Failed
  deriving (Show, Eq)

-- ============================================================================
-- 1. CONFLICT RESOLUTION
-- Higher priority wins, then higher confidence
-- ============================================================================

resolveConflict :: [Fact] -> Maybe Fact
resolveConflict [] = Nothing
resolveConflict fs = Just $ head $ sortBy cmp fs
  where
    cmp a b = compare (Down (priority a), Down (confidence a))
                      (Down (priority b), Down (confidence b))

-- Query the winning fact for a value
getBelief :: [Fact] -> String -> Maybe (Fact, Double)
getBelief facts val =
  case resolveConflict (filter (\f -> value f == val) facts) of
    Nothing -> Nothing
    Just f -> Just (f, confidence f)

-- ============================================================================
-- 2. OUTDATED DETECTION
-- Exponential decay: Conf(t) = Conf(0) * exp(-0.0001 * age_in_seconds)
-- ============================================================================

-- Half-life: time for confidence to drop to 50%
halfLifeSeconds :: Double
halfLifeSeconds = log 2.0 / 0.0001  -- ~6931 seconds

-- Compute decayed confidence at a given time
decayedConfidence :: Fact -> UTCTime -> Double
decayedConfidence f now =
  let age = realToFrac $ diffUTCTime now (timestamp f) :: Double
      decayed = confidence f * exp (-0.0001 * age)
  in decayed

-- Check if a fact is outdated (confidence < 15%)
isOutdated :: Fact -> UTCTime -> Bool
isOutdated f now = decayedConfidence f now < 0.15

-- Prune outdated facts
pruneOutdated :: [Fact] -> UTCTime -> [Fact]
pruneOutdated facts now = filter (not . flip isOutdated now) facts

-- ============================================================================
-- 3. AMBIGUOUS CONCEPTS
-- ============================================================================

-- Register a concept with multiple senses
createConcept :: String -> [String] -> Concept
createConcept nm glosses =
  Concept nm [Sense gloss Nothing 1.0 | gloss <- glosses]

-- Find senses that apply in a context
disambiguate :: Concept -> Context -> [Sense]
disambiguate c ctx =
  filter (\s -> case contextPred s of
                  Nothing -> True
                  Just pred -> pred ctx)
         (senses c)

-- Get the best (highest confidence) sense in context
bestSense :: Concept -> Context -> Maybe Sense
bestSense c ctx =
  case sortOn (Down . confidence) (disambiguate c ctx) of
    [] -> Nothing
    (s:_) -> Just s

-- ============================================================================
-- 4. MAINTAINABILITY GUARD
-- Hard limit: 80 rules per module (prevents knowledge explosion)
-- ============================================================================

-- Add a rule to a module (fails if limit exceeded)
addRule :: ModuleStats -> Rule -> Either String ModuleStats
addRule stats rule =
  if ruleCount stats >= maxRules stats
    then Left $ printf "Module %s has reached rule limit (%d rules)"
                       (LTMS.module stats)
                       (ruleCount stats)
    else Right $ stats { ruleCount = ruleCount stats + 1 }

-- Suggest refactoring when module > 70% full
needsRefactor :: ModuleStats -> Bool
needsRefactor stats =
  let percent = fromIntegral (ruleCount stats) / fromIntegral (maxRules stats) :: Double
  in percent > 0.7

-- Module complexity percentage (0-100)
complexityPercent :: ModuleStats -> Double
complexityPercent stats =
  (fromIntegral (ruleCount stats) / fromIntegral (maxRules stats)) * 100.0

-- ============================================================================
-- 5. HYBRID / NON-RULE KNOWLEDGE
-- Symbolic proof → Embedding fallback → Failure
-- ============================================================================

-- Pure symbolic prover (stub: just checks if fact exists)
proveSymbolic :: String -> [Fact] -> Maybe Fact
proveSymbolic goal facts =
  case filter (\f -> value f == goal) facts of
    [] -> Nothing
    (f:_) -> Just f

-- Embedding retrieval (stub: returns Nothing, can be extended to query vector DB)
embeddingRetrieve :: String -> Maybe (String, Double)
embeddingRetrieve _goal = Nothing  -- placeholder

-- Hybrid query: try symbolic first, then embedding
hybridQuery :: String -> [Fact] -> HybridResult Fact
hybridQuery goal facts =
  case proveSymbolic goal facts of
    Just f -> Symbolic f 1.0
    Nothing -> case embeddingRetrieve goal of
      Just (_res, conf) -> Embedding (Fact goal "embedding" (unsafePerformIO getCurrentTime) conf 0) conf
      Nothing -> Failed

-- ============================================================================
-- QUERY & ASSERTION API
-- ============================================================================

-- Get all facts for a value (before conflict resolution)
getAllFacts :: [Fact] -> String -> [Fact]
getAllFacts facts val = filter (\f -> value f == val) facts

-- Get facts from a specific source
factsBySource :: [Fact] -> String -> [Fact]
factsBySource facts src = filter (\f -> source f == src) facts

-- ============================================================================
-- SUBSUMPTION & TAXONOMIES
-- ============================================================================

-- Represent is-a relationships as facts
data Subsumption = Subsumption
  { subSpecific :: String
  , subGeneral :: String
  , subConfidence :: Double
  } deriving (Show, Eq)

-- Get all ancestors of a concept (transitive closure of is-a)
ancestors :: [Subsumption] -> String -> Set.Set String
ancestors subs concept = go concept Set.empty
  where
    go current visited
      | Set.member current visited = visited
      | otherwise =
          let visited' = Set.insert current visited
              parents = [subGeneral s | s <- subs, subSpecific s == current]
          in Set.unions $ visited' : [go p visited' | p <- parents]

-- ============================================================================
-- INSPECTION & DEBUGGING
-- ============================================================================

-- Pretty print a fact
factSummary :: Fact -> String
factSummary f =
  printf "Fact: %s\n  Source: %s\n  Confidence: %d%%\n  Priority: %d"
         (value f)
         (source f)
         (round $ confidence f * 100 :: Int)
         (priority f)

-- Pretty print a concept
conceptSummary :: Concept -> String
conceptSummary c =
  printf "Concept: %s\n  Senses: %d"
         (name c)
         (length $ senses c)

-- Module report
moduleReport :: [ModuleStats] -> String
moduleReport stats =
  unlines $
    [ printf "%s: %d/%d rules (%.1f%%)"
             (LTMS.module s) (ruleCount s) (maxRules s) (complexityPercent s)
    | s <- stats
    ] ++
    [ "" ] ++
    [ printf "Needs refactor (>70%%): %s"
             (unwords [LTMS.module s | s <- stats, needsRefactor s])
    ]

-- ============================================================================
-- TESTS
-- ============================================================================

runLTMSTests :: IO ()
runLTMSTests = do
  putStrLn "=== LTMS Tests (Haskell) ===\n"

  now <- getCurrentTime

  -- Test 1: Conflict resolution
  putStrLn "Test 1: Conflict resolution"
  let fact1 = Fact "color-car" "source-a" now 0.8 10
      fact2 = Fact "color-car" "source-b" now 0.9 5
      facts = [fact1, fact2]
  case resolveConflict facts of
    Just winner -> putStrLn $ "✓ Winner: " ++ source winner ++ " (confidence " ++ show (confidence winner) ++ ")"
    Nothing -> putStrLn "✗ No winner found"
  putStrLn ""

  -- Test 2: Outdated detection
  putStrLn "Test 2: Outdated detection"
  let oldTime = addUTCTime (-10 * 60) now  -- 10 minutes ago
      oldFact = Fact "old-value" "old-source" oldTime 0.5 1
  if isOutdated oldFact now
    then putStrLn "✓ Old fact marked as outdated"
    else putStrLn "✗ Old fact not detected"
  putStrLn ""

  -- Test 3: Concepts & disambiguation
  putStrLn "Test 3: Concepts & disambiguation"
  let bankConcept = createConcept "bank" ["financial institution", "river bank", "pile of snow"]
  putStrLn $ "✓ Bank concept has " ++ show (length $ senses bankConcept) ++ " senses"
  putStrLn ""

  -- Test 4: Module complexity
  putStrLn "Test 4: Module complexity guard"
  let stats = ModuleStats "main" 1 80 now
  putStrLn $ printf "✓ Main module: %.1f%% full" (complexityPercent stats)
  putStrLn ""

  -- Test 5: Hybrid query
  putStrLn "Test 5: Hybrid query with fallback"
  let skyFact = Fact "sky-color" "observation" now 0.95 100
      facts' = [skyFact]
  case hybridQuery "sky-color" facts' of
    Symbolic f _ -> putStrLn $ "✓ Query result: " ++ value f ++ " (symbolic)"
    Embedding f _ -> putStrLn $ "✓ Query result: " ++ value f ++ " (embedding)"
    Failed -> putStrLn "✗ Query failed"
  putStrLn ""

  putStrLn "=== LTMS Tests Complete ===\n"

-- Helper to get current time for tests
unsafePerformIO :: IO a -> a
unsafePerformIO = System.IO.Unsafe.unsafePerformIO
